pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -1654,6 +1669,24 @@ private long setQueryTimeout(String tableNameWithType, Map<String, String> query
     return remainingTimeMs;
   }
 
+  /**
+   * Sets a query option indicating the maximum response size that can be sent from a server to the broker. This size
+   * is measured for the serialized response.
+   */
+  private void setServerSerializedResponseLength(int numServers, @Nullable BrokerRequest brokerRequest) {
+    if (brokerRequest == null) {
+      return;
+    }
+
+    Map<String, String> queryOptions = brokerRequest.getPinotQuery().getQueryOptions();
+    Long maxLength = QueryOptionsUtils.getMaxSerializedResponseLengthPerServer(queryOptions);
```

Comment:
Are we introducing this as a query option or instance config or both ?

Reply:
There are 2 configs involved here:

There's a broker instance config that defines what the maximum response size for single query can be. If this is set, this total value is divided by the number of routed servers to determine the `per_server_budget`. This `per_server_budge` is set as a query config very similar to how we handle timeout today (the alternative is to make thrift changes to pass this info).